exporters/webhook/app.py

```diff
@@ -0,0 +1,202 @@
+#    a copy of the License at
+#
+#         http://www.apache.org/licenses/LICENSE-2.0
+#
+#    Unless required by applicable law or agreed to in writing, software
+#    distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+#    WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+#    License for the specific language governing permissions and limitations
+#    under the License.
+#
+
+import asyncio
+import http
+import json
+import logging
+
+from fastapi import Depends, FastAPI, Header, Request, Response
+from fastapi.responses import PlainTextResponse
+from prometheus_client import Counter, generate_latest
+from prometheus_client.core import REGISTRY
+from pydantic import BaseModel, ValidationError, parse_obj_as
+
+import pelorus
+from committime import CommitMetric
+from pelorus.config import load_and_log
+from webhook.models.github_webhook import GitHubCommit, GitHubDeliveryHeaders
+from webhook.models.in_memory_metric import (
+    PelorusGaugeMetricFamily,
+    in_memory_commit_metrics,
+)
+from webhook.models.pelorus_webhook import (
+    CommitTimePelorusPayload,
+    PelorusDeliveryHeaders,
+)
+
+pelorus.setup_logging()
+
+webhook_received = Counter("webhook_received_total", "Number of received webhooks")
+webhook_processed = Counter("webhook_processed_total", "Number of processed webhooks")
+
+commit_hash_set = set()
+
+# APIRouter here... ?
+app = FastAPI(title="Pelorus Webhook receiver")
+
+
+class PelorusWebhookResponse(BaseModel):
+    http_response: str
+    http_response_code: int
+
+
+def allowed_hosts(request: Request) -> bool:
+    # Raise exception if the request is not from allowed hosts
+    return True
+
+
+# This should be our env/secret from env
+# Plus it's dependent on the service itself. e.g. github may use different secret
+# https://towardsdev.com/build-a-webhook-endpoint-with-fastapi-d14bf1b1d55d
+SECRET_TOKEN = None
+
+
+@app.post(
+    "/pelorus/webhook",
+    status_code=http.HTTPStatus.ACCEPTED,
+    dependencies=[Depends(allowed_hosts)],
+)
+async def pelorus_webhook(
+    request: Request,
+    response: Response,
+    payload: dict,
+    user_agent: str = Header(None),
+    content_length: int = Header(...),
+) -> PelorusWebhookResponse:
+
+    webhook_received.inc()
+
+    if content_length > 100000:
+        return PelorusWebhookResponse(
+            http_response="Content length too big",
+            http_response_code=http.HTTPStatus.BAD_REQUEST,
+        )
+
+    payload_headers = None
+    payload_model = None
+
+    try:
+        logging.debug("User-agent: %s" % user_agent)
+
+        if user_agent.startswith("GitHub-Hookshot/"):
+            payload_headers = parse_obj_as(GitHubDeliveryHeaders, request.headers)
+        elif user_agent.startswith("Pelorus-Webhook/"):
+            payload_headers = parse_obj_as(PelorusDeliveryHeaders, request.headers)
+        else:
+            return PelorusWebhookResponse(
```

Comment:
when there is an error I believe you can raise http exception. [example](https://github.com/mateusoliveira43/log-api/blob/deb8ac6f8986d1f78840933c7d837db17c4c4e3d/source/database/functions/user.py#L34-L37)

Reply:
changed with appropriate http error codes. This is just PoC so thought it's nice to use pydantic to control the data structure of returned item. The pydantic doesn't allow easilly wrap the exceptions around, but raise http is also ok.

Reply:
I think you were right not using the  http exception, the way you wanted we can after do like this https://fastapi.tiangolo.com/advanced/additional-responses/ and add to the openapi `http://127.0.0.1:8000 /docs`